### PR TITLE
[graphql/rpc] support variables in client

### PR DESCRIPTION
**Pull request by oxade**
## Description 

Allow passing variables in the graphql client.
This will make testing suite much more flexible as we can swap out value easier. 

## Test Plan 

e2e test

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-graphql-rpc/src/client/mod.rs [hunk 1]
```diff
@@ -24,6 +24,20 @@ pub enum ClientError {
         usage_name: String,
         usage_value: Value,
     },
+    #[error("{item_type} at pos {idx} must not be empty")]
+    InvalidEmptyItem { item_type: String, idx: usize },
+    #[error("Variable {var_name} previously defined {var_type_prev} vs {var_type_curr}")]
```
Comment (stefan-mysten) [inline crates/sui-graphql-rpc/src/client/mod.rs:29]: This message is slightly confusing. Does this try to say that a previous variable with the same name was defined, which is not allowed?

File: crates/sui-graphql-rpc/src/client/simple_client.rs [hunk 2]
```diff
@@ -18,6 +21,50 @@ use crate::{
...
+}
+
+pub struct GraphqlVariables {
+    // A list of the variable name to its type
+    pub type_defintions: Vec<(String, String)>,
+    // Maps the variable names to the values
+    pub value_mappings: BTreeMap<String, Value>,
+}
+
+// impl GraphqlVariables {
```
Comment (stefan-mysten) [inline crates/sui-graphql-rpc/src/client/simple_client.rs:38]: Cleanup

File: crates/sui-graphql-rpc/src/client/simple_client.rs [hunk 3]
```diff
@@ -18,6 +21,50 @@ use crate::{
 };
 
 use super::response::GraphqlResponse;
+
+pub struct GraphqlVariable {
```
Comment (stefan-mysten) [inline crates/sui-graphql-rpc/src/client/simple_client.rs:25]: nit: maybe a better name is QueryVariable?

File: crates/sui-graphql-rpc/src/client/mod.rs (hunk 1 above)
Comment (oxade) [reply crates/sui-graphql-rpc/src/client/mod.rs:29]: Indeed, fixed now.

File: crates/sui-graphql-rpc/src/client/simple_client.rs (hunk 2 above)
Comment (oxade) [reply crates/sui-graphql-rpc/src/client/simple_client.rs:38]: Good callout, not sure how I missed that

File: crates/sui-graphql-rpc/src/client/simple_client.rs (hunk 3 above)
Comment (oxade) [reply crates/sui-graphql-rpc/src/client/simple_client.rs]: Sure, renamed.

**Review by stefan-mysten — APPROVED** [review]
Looks good! 
Don't forget to address the todo 
https://github.com/MystenLabs/sui/pull/14790/files#diff-ed440c1b687171c1a15f063ef8a10b0fe62ef09a9223858c6460ba8e1d23624aR100

**Pull request merged**